exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java

```diff
@@ -285,8 +648,18 @@ public AggOutcome doWork() {
       // In the future HashAggregate may also need to perform some actions conditionally
       // in the outer try block.
 
+      assert ! handlingSpills || currentIndex < Integer.MAX_VALUE;
+
       outside:
       while (true) {
+
+        // This would be called only once - after actual data arrives on incoming
+        if ( schema == null && incoming.getRecordCount() > 0 ) {
+          this.schema = incoming.getSchema();
+          // Calculate the number of partitions based on actual incoming data
+          delayedSetup();
+        }
+
```

Comment:
Annotate history points to @amansinha100 as the creator .... And the trace level logging is already used heavily for "normal monitoring" (e.g., number of partitions chosen, size estimates). Need some new levels below trace, which unfortunately cannot be created in LogBack .....